### PR TITLE
Compute/use data pruning metrics in metaseq

### PR DESCRIPTION
**Pull request by ktirumalafb**
**Patch Description**
Adding changes to log_model_outputs to compute data pruning metrics and use computed metrics to filter data while pruning

**Testing steps**
To compute metrics (metrics that are dependent on running the model pipeline), you can run something like:
```
 metaseq-train --task streaming_language_modeling /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321 --save-dir /checkpoint/ktirumala/tempdebug/ --arch transformer_lm --share-decoder-input-output-embed --dropout 0.1 --optimizer adam --adam-betas '(0.9, 0.98)' --weight-decay 0.01 --clip-norm 0.0 --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 --tokens-per-sample 512 --sample-break-mode none --tokens-per-sample 2048 --update-freq 1 --fp16 --log-format json --log-interval 1 --validate-interval-updates 1000 --save-interval-updates 100 --keep-interval-updates 1 --vocab-filename /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-vocab.json --merges-filename /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-merges.txt --valid-subset valid/redditflattened --batch-size 2 --compute-data-pruning-metrics ppl --compute-data-pruning-metrics-savedir /checkpoint/ktirumala/lm_data_pruning_v2/
```

Right now, the supported metrics for `--compute-data-pruning-metrics` are `ppl` and `ssl_prototypes` 

To use data pruning metrics, you should create a `.jsonl` file where each line is:
```
{
            "name": dataset_name, 
            "index": index in dataset jsonl file, 
            "metric": metric value
}
```

Then you can run 
```
metaseq-train --task streaming_language_modeling /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321 --save-dir /checkpoint/ktirumala/tempdebug/ --arch transformer_lm --share-decoder-input-output-embed --dropout 0.1 --optimizer adam --adam-betas '(0.9, 0.98)' --weight-decay 0.01 --clip-norm 0.0 --lr 0.0005 --lr-scheduler inverse_sqrt --warmup-updates 4000 --warmup-init-lr 1e-07 --tokens-per-sample 512 --sample-break-mode none --tokens-per-sample 2048 --update-freq 1 --fp16 --log-format json --log-interval 1 --validate-interval-updates 1000 --save-interval-updates 100 --keep-interval-updates 1 --vocab-filename /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-vocab.json --merges-filename /datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-merges.txt --valid-subset valid/redditflattened --batch-size 2 --use-data-pruning-metrics --use-data-pruning-metrics-filepath /checkpoint/ktirumala/tempdebug/metrics.jsonl
```

### CODE REVIEW

File: metaseq/data/filter_dataset.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    
+    Hard/easy to learn are defined by a metric between 0.0 and 1.0, with
+    0.0 meaning easy to learn, and 1.0 meaning hard to learn.
+
+    During initialization, this class expects:
+    - `frac_data`: how much of the original dataset we want to keep 
+    during training. It calculates the closest `frac_data` we can keep
+    (keep the hard examples, throw away easy examples, 
+    as per https://arxiv.org/abs/2206.14486.
+    - `metric_file`: path to jsonl file, where each line should be
```
Comment (ktirumalafb) [inline metaseq/data/filter_dataset.py:26]: The reason I chose the specified format for the metrics file: in the future, if we want to compute metrics that are done offline (do not involve running a metaseq model), we can still use that metric to filter data points at train time

File: metaseq/tasks/streaming_language_modeling.py [hunk 2]
```diff
@@ -308,13 +360,37 @@ def load_dataset(self, split: str, epoch=1, combine=False, **kwargs):
...
+            dataset_index_counter += 1
+
         assert len(datasets) > 0
 
         if self.args.multicorpus_sampling_alpha != 1:
             datasets = self._alpha_sampling(datasets, corpora, epoch)
 
         dataset = torch.utils.data.ConcatDataset(datasets)
 
+        # Filter by metric (only on train dataset)
```
Comment (ktirumalafb) [inline metaseq/tasks/streaming_language_modeling.py:392]: I put the filter dataset here instead of later on, because there's no point in processing on data points we are sure that we are not going to use (i.e shuffling, or dividing into token blocks);

however, we want to filter after concat because I thought it would be cleaner to apply the filter one time on the entire original dataset (instead of applying a filter multiple times on portions of the dataset i.e on the individual `JsonlDataset`'s)

File: metaseq/checkpoint_utils.py [hunk 3]
```diff
@@ -495,8 +496,8 @@ def load_model_ensemble_and_task(
 
             logger.info(f"CFG -------->>>>>>>>>>> {cfg.model}")
             
-            cfg.task.merges_filename = "/large_experiments/xlmg/data/gptz/tokenizers/gpt2-merges.txt"
-            cfg.task.vocab_filename = "/large_experiments/xlmg/data/gptz/tokenizers/gpt2-vocab.json"
+            cfg.task.merges_filename = "/datasets01/gptz_corpus_dedup_10_10_1_0.05_exp29/120321/tokenizers/gpt2-merges.txt"
```
Comment (ktirumalafb) [inline metaseq/checkpoint_utils.py]: Changing because `xlmg` partition is gone

File: metaseq/data/filter_dataset.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+from metaseq.file_io import PathManager
+
+from metaseq.data import data_utils
+from . import BaseWrapperDataset
+
+class FilterDataset(BaseWrapperDataset):
+    """Filters dataset by excluding examples that are easy/hard to learn.
+    
+    Hard/easy to learn are defined by a metric between 0.0 and 1.0, with
```
Comment (ktirumalafb) [inline metaseq/data/filter_dataset.py:18]: I took this part from the beyond neural scaling law paper - in reality, we might want to include a mix of "hard" and "easy" points (or maybe metrics are not able to even cleanly distinguish between the two); we can add more nuanced logic here to include datapoints in ranges of value instead of just going from 0.0 --> 1.0

File: metaseq/data/filter_dataset.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        assert 0.0 <= frac_data <= 1.0
+        self.frac_data = frac_data
+        self.dataset = dataset
+        self.metric_data = metric_data
+
+        # This means we should not be pruning, because not enough datapoints have a computed metric
+        limit = int(np.ceil(len(self.dataset) * self.frac_data))
+        self.length = limit
+
+        df_final.sort_values('metric', inplace=True, ascending=False)
```
Comment (danielsimig) [inline metaseq/data/filter_dataset.py]: Something is not quite right with variable naming here?

File: metaseq/tasks/streaming_language_modeling.py [hunk 6]
```diff
@@ -308,13 +360,37 @@ def load_dataset(self, split: str, epoch=1, combine=False, **kwargs):
...
+        if split == "train" and self.args.use_data_pruning_metrics:
+            # This method should only be called *once* when we 
+            # are processing the combined `train` dataset
+            metric_df = FilterDataset.retrieve_metric_df(
+                metric_file=self.args.use_data_pruning_metrics_filepath,
+                dataset_name_to_index=dataset_name_to_index
+            )
+            n_metric_df = len(metric_df)
+            if n_metric_df >= len(dataset):
```
Comment (danielsimig) [inline metaseq/tasks/streaming_language_modeling.py]: Can we do it such that we simply ignore data points if they are not in the metric table? Just because inference failed for a few data points (e.g. the current scenario) we could still run experiments.
